category_encoders/target_encoder.py

```diff
@@ -113,19 +120,33 @@ def __init__(self, verbose=0, cols=None, drop_invariant=False, return_df=True, h
                           category=FutureWarning)
         self.mapping = None
         self._mean = None
-        if hierarchy:
+        if isinstance(hierarchy, (dict, pd.DataFrame)) and cols is None:
+            raise ValueError('Hierarchy is defined but no columns are named for encoding')
+        if isinstance(hierarchy, dict):
             self.hierarchy = {}
             self.hierarchy_depth = {}
             for switch in hierarchy:
                 flattened_hierarchy = util.flatten_reverse_dict(hierarchy[switch])
-
                 hierarchy_check = self._check_dict_key_tuples(flattened_hierarchy)
                 self.hierarchy_depth[switch] = hierarchy_check[1]
                 if not hierarchy_check[0]:
                     raise ValueError('Hierarchy mapping contains different levels for key "' + switch + '"')
                 self.hierarchy[switch] = {(k if isinstance(t, tuple) else t): v for t, v in flattened_hierarchy.items() for k in t}
-        else:
+        elif isinstance(hierarchy, pd.DataFrame):
+            self.hierarchy = hierarchy
+            self.hierarchy_depth = {}
+            for col in self.cols:
+                HIER_cols = self.hierarchy.columns[self.hierarchy.columns.str.startswith(f'HIER_{col}')].values
+                HIER_levels = [int(i.replace(f'HIER_{col}_', '')) for i in HIER_cols]
+                if np.array_equal(sorted(HIER_levels), np.arange(1, max(HIER_levels)+1)):
```

Comment:
I think we should document somewhere that in case of supplying a dataframe as hierarchy that the columns have to be called HIER_col_1, ... HIER_col_N.  
Also I'm not sure if 1 or N is the coarsest level (so probably also worth documenting)